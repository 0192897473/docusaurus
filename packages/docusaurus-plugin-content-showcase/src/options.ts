/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Joi, RouteBasePathSchema} from '@docusaurus/utils-validation';
import {GlobExcludeDefault} from '@docusaurus/utils';
import type {OptionValidationContext} from '@docusaurus/types';
import type {PluginOptions, Options} from '@docusaurus/plugin-content-showcase';

export const DEFAULT_OPTIONS: PluginOptions = {
  id: 'showcase',
  path: 'src/showcase/website', // Path to data on filesystem, relative to site dir.
  routeBasePath: '/', // URL Route.
  include: ['**/*.{yml,yaml}'], // Extensions to include.
  exclude: GlobExcludeDefault,
};

const PluginOptionSchema = Joi.object<PluginOptions>({
  path: Joi.string().default(DEFAULT_OPTIONS.path),
  routeBasePath: RouteBasePathSchema.default(DEFAULT_OPTIONS.routeBasePath),
  include: Joi.array().items(Joi.string()).default(DEFAULT_OPTIONS.include),
  exclude: Joi.array().items(Joi.string()).default(DEFAULT_OPTIONS.exclude),
  id: Joi.string().default(DEFAULT_OPTIONS.id),
});

export function validateOptions({
  validate,
  options,
}: OptionValidationContext<Options, PluginOptions>): PluginOptions {
  const validatedOptions = validate(PluginOptionSchema, options);
  return validatedOptions;
}
