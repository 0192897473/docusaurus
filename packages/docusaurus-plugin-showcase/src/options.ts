/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Joi, RouteBasePathSchema} from '@docusaurus/utils-validation';
import type {OptionValidationContext} from '@docusaurus/types';
import type {PluginOptions, Options} from '@docusaurus/plugin-showcase';

export const DEFAULT_OPTIONS: PluginOptions = {
  id: 'showcase',
  path: 'src/showcase/website', // Path to data on filesystem, relative to site dir.
  routeBasePath: '/', // URL Route.
};

const PluginOptionSchema = Joi.object<PluginOptions>({
  path: Joi.string().default(DEFAULT_OPTIONS.path),
  routeBasePath: RouteBasePathSchema.default(DEFAULT_OPTIONS.routeBasePath),
});

export const contentAuthorsSchema = Joi.object({
  title: Joi.string().required(),
  description: Joi.string().required(),
  preview: Joi.string().required(),
  website: Joi.string().required(),
  source: Joi.string().required(),
  tags: Joi.array().items(Joi.string()).required(),
});

export function validateOptions({
  validate,
  options,
}: OptionValidationContext<Options, PluginOptions>): PluginOptions {
  const validatedOptions = validate(PluginOptionSchema, options);
  return validatedOptions;
}
