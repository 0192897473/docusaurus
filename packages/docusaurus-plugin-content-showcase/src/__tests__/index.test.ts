/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {loadContext} from '@docusaurus/core/src/server/site';
import {normalizePluginOptions} from '@docusaurus/utils-validation';

import pluginContentPages from '../index';
import {validateOptions} from '../options';

// todo add test with tags in config
// todo add test with tags in yaml

describe('docusaurus-plugin-content-showcase', () => {
  it('loads simple showcase', async () => {
    const siteDir = path.join(__dirname, '__fixtures__', 'website');
    const context = await loadContext({siteDir});
    const plugin = pluginContentPages(
      context,
      validateOptions({
        validate: normalizePluginOptions,
        options: {
          // todo broken because we use aliasedPaths
          path: 'src/showcase',
        },
      }),
    );
    const showcaseMetadata = await plugin.loadContent!();

    expect(showcaseMetadata).toMatchSnapshot();
  });
});
