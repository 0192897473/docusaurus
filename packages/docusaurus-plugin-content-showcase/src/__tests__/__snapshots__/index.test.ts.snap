// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docusaurus-plugin-content-showcase loads simple showcase 1`] = `
{
  "website": [
    {
      "description": "World",
      "preview": "github.com/ozakione.png",
      "source": "https://github.com/facebook/docusaurus",
      "tags": [
        "opensource",
        "meta",
      ],
      "title": "Hello",
      "website": "https://docusaurus.io/",
    },
  ],
}
`;
